[Merged by Bors] - feat: add AffineEquiv.ofLinearEquiv

jessealama opened the pull request:
Adds a constructor for building affine equivalences from linear equivalences and base points.

---

Extracted from #30854.

cc @Vierkantor @YaelDillies

jessealama committed:
feat: add AffineEquiv.ofLinearEquiv

Add a constructor for affine equivalences from linear equivalences and base points.

Extracted from #30854.

YaelDillies reviewed (commented):
Can you prove `ofLinearEquiv_refl`, `ofLinearEquiv_trans`?

YaelDillies commented on Mathlib/LinearAlgebra/AffineSpace/AffineEquiv.lean:
```diff
@@ -561,6 +561,25 @@ theorem pointReflection_fixed_iff_of_module [Invertible (2 : k)] {x y : P₁} :
+Given a linear equivalence `A : V ≃ₗ[k] V` and base points `f₀ g₀ : P`, this constructs
+the affine equivalence `T x = A (x -ᵥ f₀) +ᵥ g₀`. This is the standard way to convert
+a linear automorphism into an affine automorphism with specified base point mapping. -/
+def ofLinearEquiv {k : Type*} {V : Type*} {P : Type*}
```
```suggestion
def ofLinearEquiv {k : V P : Type*}
```
Same below. Even better would be to have these declared in a `variable` somewhere

jsm28 commented on Mathlib/LinearAlgebra/AffineSpace/AffineEquiv.lean:
```diff
@@ -561,6 +561,25 @@ theorem pointReflection_fixed_iff_of_module [Invertible (2 : k)] {x y : P₁} :
+    [DivisionRing k] [AddCommGroup V] [Module k V] [AddTorsor V P]
+    (A : V ≃ₗ[k] V) (f₀ g₀ : P) : P ≃ᵃ[k] P :=
+  let linearAsAffine : V ≃ᵃ[k] V :=
+    { toEquiv := A.toEquiv, linear := A, map_vadd' := fun p v => A.map_add v p }
```
I'd have expected a use of `A.toAffineEquiv` instead of this `let`.

jessealama committed:
Update Mathlib/LinearAlgebra/AffineSpace/AffineEquiv.lean

Co-authored-by: Yaël Dillies <yael.dillies@gmail.com>

jessealama commented:
> Can you prove `ofLinearEquiv_refl`, `ofLinearEquiv_trans`?

Yes! I'm on it.

jessealama committed:
refactor: use variable declarations and A.toAffineEquiv, add refl/trans theorems

YaelDillies commented on Mathlib/LinearAlgebra/AffineSpace/AffineEquiv.lean:
```diff
@@ -577,6 +577,43 @@ theorem coe_toAffineEquiv (e : V₁ ≃ₗ[k] V₂) : ⇑e.toAffineEquiv = e :=
+
+theorem ofLinearEquiv_trans (A B : V ≃ₗ[k] V) (p₀ p₁ p₂ : P) :
+    (ofLinearEquiv A p₀ p₁).trans (ofLinearEquiv B p₁ p₂) =
+      ofLinearEquiv (A.trans B) p₀ p₂ := by
```
```suggestion
@[simp]
theorem ofLinearEquiv_trans_ofLinearEquiv (A B : V ≃ₗ[k] V) (p₀ p₁ p₂ : P) :
    (ofLinearEquiv A p₀ p₁).trans (ofLinearEquiv B p₁ p₂) =
      ofLinearEquiv (A.trans B) p₀ p₂ := by
```

jessealama committed:
Update Mathlib/LinearAlgebra/AffineSpace/AffineEquiv.lean

Co-authored-by: Yaël Dillies <yael.dillies@gmail.com>

jessealama committed:
Update Mathlib/LinearAlgebra/AffineSpace/AffineEquiv.lean

Co-authored-by: Yaël Dillies <yael.dillies@gmail.com>

jessealama committed:
refactor: weaken DivisionRing to Ring in ofLinearEquiv

Addresses review comment by jsm28 at https://github.com/leanprover-community/mathlib4/pull/30950#discussion_r2478269916

jessealama committed:
refactor: rename params and simplify proofs in ofLinearEquiv

YaelDillies reviewed (approved):
Thanks!

maintainer delegate

jessealama committed:
Update Mathlib/LinearAlgebra/AffineSpace/AffineEquiv.lean

Co-authored-by: Yaël Yanis Dillies <yael.dillies@gmail.com>

jessealama commented on Mathlib/LinearAlgebra/AffineSpace/AffineEquiv.lean:
```diff
@@ -577,6 +577,48 @@ theorem coe_toAffineEquiv (e : V₁ ≃ₗ[k] V₂) : ⇑e.toAffineEquiv = e :=
+
+@[simp]
+theorem ofLinearEquiv_linear (A : V ≃ₗ[k] V) (p₀ p₁ : P) :
+    (ofLinearEquiv A p₀ p₁).linear = A :=
```
Thanks! Sometimes I'm not 100% clear on what to call things; any feedback is appreciated.

j-loreaux commented on Mathlib/LinearAlgebra/AffineSpace/AffineEquiv.lean:
```diff
@@ -577,6 +577,48 @@ theorem coe_toAffineEquiv (e : V₁ ≃ₗ[k] V₂) : ⇑e.toAffineEquiv = e :=
+@[simp]
+theorem ofLinearEquiv_trans_ofLinearEquiv (A B : V ≃ₗ[k] V) (p₀ p₁ p₂ : P) :
+    (ofLinearEquiv A p₀ p₁).trans (ofLinearEquiv B p₁ p₂) =
+      ofLinearEquiv (A.trans B) p₀ p₂ := by
```
our `trans` lemmas go in the reverse order and it would be called `ofLinearEquiv_trans`.

j-loreaux replied:
I see. Okay, the way you had it is fine then.

jessealama replied:
OK, I'll revert the change. Thanks for taking a look.

j-loreaux reviewed (commented):
please fix, then merge (but wait for CI to finish first).

bors d+

jessealama committed:
refactor: rename ofLinearEquiv_trans_ofLinearEquiv to ofLinearEquiv_trans

Follow naming convention for trans lemmas as suggested by @j-loreaux.

jessealama commented:
> please fix, then merge (but wait for CI to finish first).

Thanks! I'll take a look at the CI in the morning and merge if it's OK. Thanks for your help with this cc @YaelDillies @jsm28

jessealama committed:
refactor: restore ofLinearEquiv_trans_ofLinearEquiv naming

The longer name `ofLinearEquiv_trans_ofLinearEquiv` better describes the lemma
structure (composing two `ofLinearEquiv` applications) compared to the shorter
`ofLinearEquiv_trans`, which could be misinterpreted as relating to a single
`ofLinearEquiv` application.

After discussion, the reviewer confirmed the original naming was appropriate.
See: https://github.com/leanprover-community/mathlib4/pull/30950#discussion_r2529571951
